wicketd/src/artifacts/extracted_artifacts.rs

```diff
@@ -0,0 +1,176 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use super::RepositoryError;
+use anyhow::Context;
+use bytes::Bytes;
+use camino::Utf8PathBuf;
+use camino_tempfile::Utf8TempDir;
+use omicron_common::update::ArtifactHash;
+use omicron_common::update::ArtifactHashId;
+use omicron_common::update::ArtifactKind;
+use sha2::Digest;
+use sha2::Sha256;
+use slog::info;
+use slog::Logger;
+use std::fs::File;
+use std::io;
+use std::io::BufWriter;
+use std::io::Read;
+use std::io::Write;
+use std::sync::Arc;
+use tokio::io::AsyncRead;
+use tokio_util::io::ReaderStream;
+
+/// Handle to the data of an extracted artifact.
+///
+/// This does not contain the actual data; use `reader_stream()` to get a new
+/// handle to the underlying file to read it on demand.
+#[derive(Debug, Clone)]
+pub(crate) struct ExtractedArtifactDataHandle {
+    tempdir: Arc<Utf8TempDir>,
+    file_size: usize,
+    hash_id: ArtifactHashId,
+}
```

Comment:
Valid concern and I'd be happy to add more documentation around the behavior, but I'm not sure about changing `Clone`. In some sense the degenerate case you describe is exactly how this is actually used: we put these into a long-lived btree map (inside an `UpdatePlan`) until a new repo is uploaded, and we want the directory to continue to exist as long as the plan is live.

Reply:
Added a couple doc comments in 96831142. I did take a stab at making this not `Clone` and having an explicit `clone_handle()` method, and it's relatively painful - we can't use `Option::cloned` anymore, and we have to either add a similar method to `ArtifactIdData` or manually implement `Clone` on it (which seems like it would kinda defeat the purpose).

I did check by hand that after uploading a second TUF repo, the temp directory created for the first is removed, so we're not unintentionally keeping that directory alive today. It would be nice to add a test for that; I'll give that a shot before landing this.